### DISCUSSION TITLE
Orchestration, deployment, monitoring etc.

### DISCUSSION THREAD

**Discussion by leonpolak** [metarhia/Roadmap#10] [Ideas]
Для быстроты напишу на русском, потом могу заменить, если против правил.
По результатам сегодняшней встречи, мы пришли к выводу, что нужно собрать в одном месте все соображения по подготовке Метархии для быстрого развёртывания, управления, мониторинга и скоординированной работы в распределённых многосерверных системах.

Я не вижу сейчас возможности как следует сделать серьёзную систему взамен применения существующих и проверенных бенчмарками механизмов, типа Докера или Подмена для контейнеров и Кубернетиса или Номада для оркестрации. Да, было-бы приятно иметь в наличии своё кастомное простое решение, но, сейчас не хватит ни рук, ни опыта, ни денег, ни времени. В конце концов, те потребители, которым нужна будет мультиконтейнерная среда, смогут позволить себе девопса и настройку существующих инструментов. На сегодняшний день, как мне кажется, это единственный серьёзный вариант, в конце концов, большие клиенты, которым будут нужны сотни или тысячи инстансов, в дверь пока не ломятся с мешками денег и появятся не сразу.

Вместо этого, я предлагаю сейчас, после решения супер срочных проблем, заняться механизмами сбора специфических для Метрахии метрик, типа выделенных ресурсов и загрузки тредов, и налаживания связи по метакому с управляющими центральными серверами, куда они могут отсылать метрики и откуда они смогут получать команды  (список собираемых метрик и команд нужно будет здесь описать списком). К этим управляющим серверам я могу сделать админский дашбоард и через него отображать сохранённые в базе собранные метрики и пересылать команды от оператора. 

Начать можно с примера, когда один сервер будет и управляющим и исполняющим, то есть отсылать метрики сам-себе и принимать команды сам от себя. Когда будет готова и протестировано межсерверное взаимодействие инстансов в кластерах, адреса для связи и механизмы аутентикации можно будет просто поменять в конфиге.

Кроме этого, нужны будут примеры контейниризированных инстансов метархии управляемых существующими механизмами, как минимум для того чтобы показать как это делается потенциальным клиентам. Сделать их нужно несколько, с применением наиболее популярных инструментов, как я уже выше написал: Кубернетис, Номад, Докер, Подмен, Терраформ и Анзибль.

Поэтому, команде оркестрации предлагаю заняться этими двумя задачам: 
1) административный/исполняющий сервер
2) примеры использования существующих инструментов

**Comment by leonpolak** [comment]
Почитал я тут и посмотрел материалы, и, надо сказать, что open source продукты HashiCorp: Packer, Vagrant, Terraform, Consul, Boundary, Vault, Nomad, and Waypoint, довольно просты в использовании, бесплатны и покрывают практически все аспекты развёртывания и контроля распределённой инфраструктуры. Если как следует научиться ими пользоваться и настроить на использование Метархии, то проблемы, как мне кажется, все решаться, особенно на этом этапе.

**Reply by MarhiievHE** [reply]
@leonpolak На сколько я помню, в дискашенах разрешены все языки) Из того что вы перечислили в комментарии, я только немного Terraform знаю, приходилось немного под aws настраивать инфраструктуру с его помощью. Но как вы написали выше, обычно те кому нужно такое большое количество инстансов и автоматизация уже обычно могут позволить себе иметь в штате хорошего девопса, а он уже и без нас знает как работать с облачными сервисами и знаком с автоматизацией. 
Я вот что думал, мы могли бы сделать для начала вариант который мог бы к примеру упростить жизнь тем же стартапам, чтобы они могли относительно быстро и удобно разработать прототип на базе метархии и развернуть его на одном или нескольких серверах / виртуалках.

**Comment by tshemsedinov** [comment]
Предлагаю план
- Сбор метрик памяти, cpu, i/o из тредов через `perf_hooks` и `worker_threads.performance` https://nodejs.org/api/worker_threads.html#workerperformance
- Консолидация метрик (объединение их за периоды и статистика)
- Логирование метрик в файлы и/или SaaS решение по API
- Автоматически забирать изменения в master (или указанной) ветке репозитория проекта по web-хуку от гитхаба, когда выходит новый релиз, чтобы impress подтянул эти изменения.
- Простое решение доя докер-контейнера для быстрого развертывания приложения метархии на стенделон сервере

**Reply by MarhiievHE** [reply]
Неплохо было бы еще собирать например метрику по количеству обработанных запросов к апи, но то такое.
Я так понимаю мы подразумеваем нашу систему управления-мониторинга как отдельный сервис? Тогда можно логи метрик в базу писать как и другие системы мониторинга делают, из этих данных можно будет потом через апи графики на фронте строить и оповещения в телегу к примеру отправлять
Ну если совсем простое для одного сервера то можно просто генерировать компосе для начала. Меня вот только смущает, кто наш сервис то разворачивать будет?)

**Reply by leonpolak** [reply]
"кто наш сервис то разворачивать будет"? 
После того, как мы Metactos в прадакшен запустим, мы сами - для подписчиков сервиса, для которого у нас уже довольно много потенциальных кандидатов и есть более-менее чёткий маркетинговый план. Если до этого (или после) придут заказчики для кастомного standalone решения - они с нашей помощью, по индивидуальному контракту.
Но, для этого нужен кто-то, кто поизучает вопрос,  поэкспериментирует и приспособит под Метархию перечисленные инструменты (или наоборот) и сделает примеры их использования. Тогда бы, на основе этих примеров, я мог сделать инструменты управления этими вещами прямо из IDE, например, полдуавтоматически формировать на фронте конфигурацию, удалённо релактировать эти их YAML файлы и перезапускать удалённо их сервисы на машинах, Ну, или, как минимум были-бы проверенные  апробированные решения по развёртыванию, что сильно сэкономило-бы время простым копи-пэйстом.

**Comment by leonpolak** [comment]
Тимур, насчёт плана:
1) Я более, чем уверен, что существующая система логировани в ноде позволяет туда записывать любые кастомные метрики и вообще хоть чёрта лысого, и потом их фильтровать и вытаскивать существующими системами мониторинга, так что предлагаю пока попробовать это наиболее простое решение, 
2) Было бы очень здорово, чтобы в этих кастомных метриках фигурировали tenantID, чтобы можно было тенантам показывать только относящиеся к их тредам логи
3) Использование существующих систем для считывани логов - решение временное, но позволяющее быстро запустить всю тему. Темм не менее, надо как следует поизучать вопрос и руками потыкать - что нам подходит.

**Reply by tshemsedinov** [reply]
В ноде нет системы логирования, у нас свой metalog и он сразу пишет отдельные логи для каждого треда, так что, туда и тенантId не нужно писать. Просто пишем сериализованную струкуктуру с каким-то маркером, что это метрики, чтобы отличить от всех других записей с логе, файл привязан к треду, тред к тенанту. Возможно, будет удобно имя файла к тенантId привызывать.

**Reply by leonpolak** [reply]
О как, не обращал внимания, всегда node пользовался через PM2 и думал, что это своё нодовское. Ну, если эта проблема решена через metalog, осталось только продумать систему как их удалённо получать для мониторинга (фильтруя по параметрам запроса), и какие отсылать команды на исполнение с управляющего сервера (и вообще взаимодействие мастер-слэйв).
